sklearn/impute/tests/test_impute.py

```diff
@@ -1786,3 +1786,57 @@ def test_simple_imputer_constant_fill_value_casting():
         )
         X_trans = imputer.fit_transform(X_float32)
         assert X_trans.dtype == X_float32.dtype
+
+
+@pytest.mark.parametrize("strategy", ["mean", "median", "most_frequent", "constant"])
+def test_iterative_imputer_no_empty_features(strategy):
+    """Check the behavior of the iterative imputer.
+    When the input matrix contains no empty features, it should return
+    the same result regardless of whether `keep_empty_features`.
+    """
+    rng = np.random.RandomState(0)
+    X = _sparse_random_matrix(10, 10, density=0.95, random_state=rng).toarray()
+    X[X == 0] = np.nan
+
+    imputer_false = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=False
+    )
+    X_imputed_false = imputer_false.fit_transform(X)
+
+    imputer_true = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=True
+    )
+    X_imputed_true = imputer_true.fit_transform(X)
+
+    assert_array_equal(X_imputed_false, X_imputed_true)
+
+
+@pytest.mark.parametrize("strategy", ["mean", "median", "most_frequent", "constant"])
+def test_iterative_imputer_with_empty_features(strategy):
+    """Check the behavior of the iterative imputer.
+    When the input matrix contains empty features, it should return
+    the same values for the non-empty features regardless of
+    `keep_empty_features`. Additionally, the empty features
+    should become 0s when `keep_empty_features` is set to True.
+    """
+    rng = np.random.RandomState(0)
+    X = _sparse_random_matrix(10, 10, density=0.95, random_state=rng).toarray()
+    X[X == 0] = np.nan
+    X[:, 0] = np.nan
+
+    imputer_false = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=False
+    )
+    X_imputed_false = imputer_false.fit_transform(X)
+
+    imputer_true = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=True
+    )
+    X_imputed_true = imputer_true.fit_transform(X)
+
+    if strategy == "constant":
```

Comment:
We will have different shapes of `X` depending on `initial_strategy` and `keep_empty_features`, right?

For example using this data
```python
    X = np.array(
        [
            [np.nan, np.nan, 0, 1],
            [np.nan, 2, np.nan, 3],
            [np.nan, 4, 5, np.nan]
        ]
    )
```

**Both `X`s have the same shape when initial_strategy = constant**
- keep_empty_features == False. X_imputed.shape == (3, 4)
- keep_empty_features == True. X_imputed.shape == (3, 4)

In this case we just need one test
```python
assert_allclose(X_drop_empty_features, X_keep_empty_features, rtol=1e-4)
```

**But `X`s will have different shapes when initial_strategy != constant**
- keep_empty_features == False. X_imputed.shape == (3, 3)
- keep_empty_features == True. X_imputed.shape == (3, 4)

I guess in this case we still need to check whether strategy is constant or not, then continue to test equality on non-empty features. What do you think?

Reply:
Not if you take this fix: https://github.com/scikit-learn/scikit-learn/pull/29779#discussion_r1752191625

It will drop the feature even with the `constant` strategy.

Reply:
> I see. So, the expected condition is to drop the empty features, even with constant strategy.

Yes we should be consistent. In `SimpleImputer`, it was for backward compatibility but I think that we should deprecate and so I open an issue.

Reply:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/3d04df9fdb51b8c8f20e0cf1e45860470171c82b